app/src/main/java/clarkson/ee408/tictactoev4/MainActivity.java

```diff
@@ -137,6 +139,123 @@ public void sendMove(int move) {
         });
     }
 
+    /**
+     * Sends ABORT_GAME request to server when user leaves an ongoing game
+     */
+    private void abortGame() {
+        Log.d("MainActivity", "Sending ABORT_GAME request");
+
+        // Create a Request object with type ABORT_GAME
+        Request request = new Request();
+        request.setType(RequestType.ABORT_GAME);
+        request.setData(""); // No additional data needed
+
+        // Send request asynchronously using AppExecutors
+        AppExecutors.getInstance().networkIO().execute(() -> {
+            try {
+                GamingResponse response = socketClient.sendRequest(request, GamingResponse.class);
+
+                // Process response in main thread to show Toast
+                AppExecutors.getInstance().mainThread().execute(() -> {
+                    if (response != null && response.getStatus() == ResponseStatus.SUCCESS) {
+                        Toast.makeText(MainActivity.this,
+                                "Game aborted successfully",
+                                Toast.LENGTH_SHORT).show();
+                        Log.d("MainActivity", "Game aborted successfully");
+                    } else {
+                        String errorMsg = "Failed to abort game";
+                        if (response != null && response.getMessage() != null) {
+                            errorMsg = response.getMessage();
+                        }
+                        Toast.makeText(MainActivity.this,
+                                errorMsg,
+                                Toast.LENGTH_SHORT).show();
+                        Log.e("MainActivity", errorMsg);
+                    }
+                });
+            } catch (Exception e) {
+                Log.e("MainActivity", "Error aborting game", e);
+                // Show Toast in main thread
+                AppExecutors.getInstance().mainThread().execute(() -> {
+                    Toast.makeText(MainActivity.this,
+                            "Error connecting to server",
+                            Toast.LENGTH_SHORT).show();
+                });
+            }
+        });
+    }
+
+    /**
+     * Sends COMPLETE_GAME request to server when user leaves after game completion
+     */
+    private void completeGame() {
+        Log.d("MainActivity", "Sending COMPLETE_GAME request");
+
+        // Create a Request object with type COMPLETE_GAME
+        Request request = new Request();
+        request.setType(RequestType.COMPLETE_GAME);
+        request.setData(""); // No additional data needed
+
+        // Send request asynchronously using AppExecutors
+        AppExecutors.getInstance().networkIO().execute(() -> {
+            try {
+                GamingResponse response = socketClient.sendRequest(request, GamingResponse.class);
```

Comment:
Complete game returns a standard Response

```suggestion
                Response response = socketClient.sendRequest(request, Response.class);
```